controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -757,10 +757,15 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
 		return actionError{err}
 	}
 	switch info.host.Status.Provisioning.State {
-	case metal3v1alpha1.StateRegistering, metal3v1alpha1.StateExternallyProvisioned:
+	case metal3v1alpha1.StateRegistering, metal3v1alpha1.StateExternallyProvisioned, metal3v1alpha1.StateDeleting:
 		// No need to create PreprovisioningImage if host is not yet registered
 		// or is externally provisioned
 		preprovImgFormats = nil
+	case metal3v1alpha1.StateDeprovisioning:
+		// PreprovisioningImage is not required for deprovisioning when cleaning is disabled
+		if info.host.Spec.AutomatedCleaningMode == metal3v1alpha1.CleaningModeDisabled {
```

Comment:
I don't love that we're encoding an assumption about the implementation of Ironic (i.e. that it doesn't need a PreprovisioningImage when cleaning is disabled) into the controller (rather than the ironic provisioner), but the alternative of passing the state and cleaning mode to PreprovisioningImageFormats() doesn't make a lot of conceptual sense either, so I think this is probably the best hack for now.

Reply:
I'm afraid we already have way too many assumptions about Ironic baked everywhere, including the API.